### PR TITLE
NEP: Array function protocol

### PR DESCRIPTION
**Pull request by mrocklin**
Written with Stephan Hoyer
after discussion with Stefan Van der Walt, Charles Harris, Matti Picus, Jaime Frio and Nathaniel Smith

### CODE REVIEW

File: doc/neps/nep-0016-array-function-protocol.rst [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+            return HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+    HANDLED_FUNCTIONS = {
+        np.concatenate: my_concatenate,
+        np.broadcast_to: my_broadcast_to,
+        np.sum: my_sum,
+        ...
+    }
+
+*Note from MR: would it make sense instead for them to return the
```
Comment (shoyer) [inline doc/neps/nep-0016-array-function-protocol.rst]: delete? :)

File: doc/neps/nep-0016-array-function-protocol.rst [hunk 2]
```diff
@@ -0,0 +1,517 @@
+==================================================
+NEP: Dispatch Mechanism for NumPy's high level API
+==================================================
+
+:Author: Authors: Stephan Hoyer <shoyer@google.com> and Matthew Rocklin <mrocklin@gmail.com>
```
Comment (shoyer) [inline doc/neps/nep-0016-array-function-protocol.rst]: I think the standard is one line/`:Author:` entry per author

**Review by shoyer — COMMENTED** [review]
Can we save NEP-16 for Nathaniel's `An abstract base class for identifying "duck arrays"` in #10706 ? That would make this NEP-17.

**Comment by mrocklin** [comment]
Thanks.  Done

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+protocol. Let us consider a few examples of NumPy functions and how they
+might be affected by this change:
+
+.. code:: python
+
+    def broadcast_to(array, shape, subok=False):
+        success, value = do_array_function_dance(
+            func=broadcast_to,
+            relevant_arguments=[array],
+            array, shape, subok=subok)  # *args, **kwargs
```
Comment (mattip) [inline doc/neps/nep-00XX-array-function-protocol.rst]: This does not match the function signature `(self, func, overload_types, *args, **kwargs)` and mixes positional and keyword arguments. In particular, I don't understand what is input as the `overload_types` positional argument.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+
+Coersion to a NumPy array as a catch-all fallback
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With the current design, classes that implement ``__array_function__``
+to overload at least one function implicitly declare an intent to
+implement the entire NumPy API. It's not possible to implement *only*
+``np.concatenate()`` on a type, but fall back to NumPy's default
+behavior of casting with ``np.asarray()`` for all other functions.
+
```
Comment (mattip) [inline doc/neps/nep-00XX-array-function-protocol.rst]: ?? If their implementation of `__array_function__` returns `NotImplemented` for all other functions, haven't they opted out of those functions?

**Comment by mattip** [comment]
This should be merged as nep-0018 once the initial comments are handled. Deeper discussion, as I understand the process, should take place on the mailing list

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+        return np.sum(np.exp(y))
+
+Some of this is possible today with various protocol mechanisms within
+Numpy.
+
+-  The ``np.exp`` function checks the ``__array_ufunc__`` protocol
+-  The ``.T`` method works using Python's method dispatch
+-  The ``np.sum`` function explicitly checks for a ``.sum`` method on
+   the argument
+
```
Comment (hameerabbasi) [inline doc/neps/nep-00XX-array-function-protocol.rst]: Might be worth noting here that 

- `np.sum` falls back to `__array_ufunc__` with `method='reduce'` if the object does not have a `sum` method.
- `ufunc.reduce` uses `__array_ufunc__` with `method='reduce'` by default.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+Specialized protocols
+~~~~~~~~~~~~~~~~~~~~~
+
+We could (and should) continue to develop protocols like
+``__array_ufunc__`` for cohesive subsets of Numpy functionality.
+
+As mentioned above, if this means that some functions that we overload
+with ``__array_function__`` should switch to a new protocol instead,
+that is explicitly OK for as long as ``__array_ufunc__`` retains its
```
[the remark this answers is not in the document]
Comment (hameerabbasi) [reply doc/neps/nep-00XX-array-function-protocol.rst]: Yes. In particular, it might support `gufunc`s as well as `ufunc`s in the future.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+
+The interface
+~~~~~~~~~~~~~
+
+We propose the following signature for implementations of
+``__array_function__``:
+
+.. code-block:: python
+
+    def __array_function__(self, func, types, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (shoyer) [reply doc/neps/nep-00XX-array-function-protocol.rst]: Agreed, I think we need to switch this to not use `*` and **` unpacking.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+of these may decide that, given the available inputs, they are unable to
+determine the correct result. How do we call the right one? If several
+are valid then which has precedence?
+
+The rules for dispatch with ``__array_function__`` match those for
+``__array_ufunc__`` (see
+`NEP-13 <http://www.numpy.org/neps/nep-0013-ufunc-overrides.html>`__).
+In particular:
+
+-  NumPy will gather implementations of ``__array_functions__`` from all
```
[the remark this answers is not in the document]
Comment (shoyer) [reply doc/neps/nep-00XX-array-function-protocol.rst]: No, this is a mistake

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+   specified inputs and call them in order: subclasses before
+   superclasses, and otherwise left to right.
```
[the remark this answers is not in the document]
Comment (shoyer) [reply doc/neps/nep-00XX-array-function-protocol.rst]: https://bugs.python.org/issue30140 was linked to in NEP-13

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+within every relevant Numpy function. This is a pervasive change, but of
+fairly simple and innocuous code that should complete quickly and
+without effect if no arguments implement the ``__array_function__``
+protocol. Let us consider a few examples of NumPy functions and how they
+might be affected by this change:
+
+.. code:: python
+
+    def broadcast_to(array, shape, subok=False):
+        success, value = do_array_function_dance(
```
[the remark this answers is not in the document]
Comment (shoyer) [reply doc/neps/nep-00XX-array-function-protocol.rst]: Yes, if we only supported Python 3. But in Python 2, we would lose introspection of function arguments.

Nonetheless this should indeed be mentioned.

File: doc/neps/nep-00XX-array-function-protocol.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (shoyer) [reply doc/neps/nep-00XX-array-function-protocol.rst]: I think this is probably a typo and should be `__array_function__` instead

File: doc/neps/nep-00XX-array-function-protocol.rst (hunk 5 above)
Comment (shoyer) [reply doc/neps/nep-00XX-array-function-protocol.rst]: `np.mean()` might be a better example, I'm pretty sure it doesn't do the `__array_ufunc__` fallback.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+libraries that consume the NumPy API such as xarray.
+
+
+Discussion
+----------
+
+Various alternatives to this proposal were discussed in a few Github issues:
+
+1.  `pydata/sparse #1 <https://github.com/pydata/sparse/issues/1>`_
+2.  `numpy/numpy #11128 <https://github.com/numpy/numpy/issues/11129>`_
```
Comment (hameerabbasi) [inline doc/neps/nep-00XX-array-function-protocol.rst]: Typo? Text says 11128 but issue is 11129.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 12]
```diff
@@ -213,7 +215,8 @@ might be affected by this change:
         success, value = do_array_function_dance(
             func=broadcast_to,
             relevant_arguments=[array],
-            array, shape, subok=subok)  # *args, **kwargs
+            args=(array,),
+            kwargs=dict(shape=kwargs, subok=subok))
```
Comment (hameerabbasi) [inline doc/neps/nep-00XX-array-function-protocol.rst]: Typo? Maybe should be `shape=shape`

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 13]
```diff
@@ -0,0 +1,538 @@
...
+might be affected by this change:
+
+.. code:: python
+
+    def broadcast_to(array, shape, subok=False):
+        success, value = do_array_function_dance(
+            func=broadcast_to,
+            relevant_arguments=[array],
+            args=(array,),
+            kwargs=dict(shape=kwargs, subok=subok))
```
Comment (hameerabbasi) [inline doc/neps/nep-00XX-array-function-protocol.rst]: Still a typo. `shape=shape`?

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 14]
```diff
@@ -0,0 +1,538 @@
...
+
+Example for a project implementing the NumPy API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Most implementations of ``__array_function__`` will start with two
+checks: 1. Is the given function something that we know how to overload?
+2. Are all arguments of a type that we know how to handle?
+
+If these conditions hold, ``__array_function__`` should return
+implementation for ``func(*args, **kwargs)``. Otherwise, it should
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: Given the example below I think you mean "the result from calling its implementation of ...".

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 15]
```diff
@@ -0,0 +1,538 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This will require two changes within the Numpy codebase:
+
+1. A function to inspect available inputs, look for the
+   ``__array_function__`` attribute on those inputs, and call those
+   methods appropriately until one succeeds.
+
+   This is one additional function of moderate complexity.
+2. Calling this function within all relevant Numpy functions.
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: insert empty line before this one.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 16]
```diff
@@ -0,0 +1,538 @@
...
+   function call that will be checked for an ``__array_function__``
+   implementation.
+-  The tuple ``args`` and dict ``**kwargs`` are directly passed on from the
+   original call.
+
+Unlike ``__array_ufunc__``, there are no high-level guarantees about the
+type of ``func``, or about which of ``args`` and ``kwargs`` may contain
+an objects implementing the array API. As a convenience for
+``__array_function__`` implementors of the NumPy API, ``types`` contains
+a list of all types that NumPy checks for potential overloads.
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: It is unclear what this means. A list of types of all arguments checked? A list of types of arguments that defined `__array_function``? More in the general comments.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 17]
```diff
@@ -0,0 +1,538 @@
...
+Necessary changes within the Numpy codebase itself
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This will require two changes within the Numpy codebase:
+
+1. A function to inspect available inputs, look for the
+   ``__array_function__`` attribute on those inputs, and call those
+   methods appropriately until one succeeds.
+
+   This is one additional function of moderate complexity.
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: Maybe comment here that speed will be of the essence.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 18]
```diff
@@ -0,0 +1,538 @@
...
+
+Unlike ``__array_ufunc__``, there are no high-level guarantees about the
+type of ``func``, or about which of ``args`` and ``kwargs`` may contain
+an objects implementing the array API. As a convenience for
+``__array_function__`` implementors of the NumPy API, ``types`` contains
+a list of all types that NumPy checks for potential overloads.
+
+Still be determined: what guarantees can we offer for ``types``? Should
+we promise that types are unique, and appear in the order in which they
+are checked?
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: Another question is whether it should even be here. Since a `relevant_arguments` is being considered, perhaps a `relevant_types` in `kwargs`?

File: doc/neps/nep-00XX-array-function-protocol.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply doc/neps/nep-00XX-array-function-protocol.rst]: This example really does not look nice!! If this cannot be done more elegantly in python2, then I think we should aim for python3 only - this process is unlikely to be that fast and even if it were, skipping one release would be well worth having a much simpler interface. In particular, in python3 one could use a decorator that looks at annotations (which we'd like anyway!) - much more elegant.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 19]
```diff
@@ -0,0 +1,538 @@
...
+    def broadcast_to(array, shape, subok=False):
+        success, value = do_array_function_dance(
+            func=broadcast_to,
+            relevant_arguments=[array],
+            args=(array,),
+            kwargs=dict(shape=shape, subok=subok))
+        if success:
+            return value
+
+        ... # continue with the definition of broadcast_to
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: I actually think this is slightly the wrong model: `ndarray` should define `__array_function__` itself and that should call the actual definition of `broadcast_to` (but at lowest priority). See more general comment.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 20]
```diff
@@ -0,0 +1,538 @@
...
+One downside of passing on all arguments directly on to
+``__array_function__`` is that it makes it hard to extend the signatures
+of overloaded NumPy functions with new arguments, because adding even an
+optional keyword argument would break existing overloads.
+
+This is not a new problem for NumPy. NumPy has occasionally changed the
+signature for functions in the past, including functions like
+``numpy.sum`` which support overloads.
+
+For adding new keyword arguments that do not change default behavior, we
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: I think this is too restraining. I think it would be up to implementations to check that there are no excess arguments.

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 21]
```diff
@@ -0,0 +1,538 @@
...
+arrays (CuPy), Sparse arrays (scipy.sparse, pydata/sparse) and parallel
+arrays (Dask array) as well as various Numpy-like implementations in the
+deep learning frameworks, like TensorFlow and PyTorch.
+
+Similarly there are several projects that build on top of the Numpy API
+for labeled and indexed arrays (XArray), automatic differentation
+(Autograd, Tangent), higher order array factorizations (TensorLy), etc.
+that add additional functionality on top of the Numpy API.
+
+We would like to be able to use these libraries together, for example we
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: Add:
```
Finally, some ``ndarray`` subclasses add new behaviour (e.g., MaskedArray and astropy's Quantity),
which gives different meaning to functions. Indeed, MaskedArray reimplements some of the core
numpy functions, and Quantity has long-standing issues about compatibility with stacking functions.
```

File: doc/neps/nep-00XX-array-function-protocol.rst [hunk 22]
```diff
@@ -0,0 +1,538 @@
...
+that assuredly many pandas users rely on. If pandas implemented
+``__array_function__`` only for ``np.concatenate``, unrelated NumPy
+functions like ``np.nanmean`` would suddenly break on pandas objects by
+raising TypeError.
+
+With ``__array_ufunc__``, it's possible to alleviate this concern by
+casting all arguments to numpy arrays and re-calling the ufunc, but the
+heterogeneous function signatures supported by ``__array_function__``
+make it impossible to implement this generic fallback behavior for
+``__array_function__``.
```
Comment (mhvk) [inline doc/neps/nep-00XX-array-function-protocol.rst]: This could be solved by all functions also gaining a `coerce` (or so) keyword argument, somewhat equivalent to the current `subok`, but always `False` by default.

Alternatively, as I'd like an `ndarray.__array_function__` implementation anyway (for super calls), *that* implementation could take a `coerce` keyword argument (now default `True`) or re-use `subok` (then default `False`).

**Review by mhvk — COMMENTED** [review]
The nature of types is really rather unclear. Should it perhaps be a list of types the `ndarray` implementation does not know what to do with? (This would include `ndarray` subclasses if the current function would do `asarray`)? More comments on mailing list.

**Comment by mhvk** [comment]
I cannot find this mentioned on the mailing list, so then here:

1. I'm rather unclear about the use of `types`. It can help me decide what to do, but I would still have to find the argument in question. If we pass anything at all, should it just be the argument itself that does not get immediately recognized by `ndarray`? Or, better still, perhaps a tuple of all arguments that were inspected?
2. For subclasses, it would be very handy to have `ndarray.__array_function__`, so one can call super after changing arguments. (For `__array_ufunc__`, there was lots of question about whether this was useful, but it really is!!).
3. This `ndarray.__array_function__` might also help solve the problem of cases where coercion is fine: it could have an extra keyword argument (say `coerce`) that would call the function with coercion in place.
4. Indeed, the `ndarray.__array_function__` could just be used inside the "dance" function, and then the actual implementation of a given function would just be a separate, private one.

**Comment by shoyer** [comment]
We still need to send this out to the mailing this. I'll do this this
afternoon unless Matt beats me to it.

File: doc/neps/nep-00XX-array-function-protocol.rst (hunk 17 above)
Comment (mhvk) [reply doc/neps/nep-00XX-array-function-protocol.rst]: In particular, the speed for *regular* arrays should not be impacted much even for small arrays.

**Comment by mhvk** [comment]
Another general comment, since this is still not on the mailing list, yet I'm thinking about it now: Since speed for *normal* operation should be impacted as minimally as possible, there should be obvious ways to ensure no type checking dance is done. Some possible solutions (which I think should be in the NEP, even if as discounted options):
- Two namespaces, one for the undecorated base functions, and one for the decorated ones. The idea would be that if one knows one is dealing with arrays only, one would do `import numpy.array_only as np`. The latter namespace would be the one automatically used for `ndarray.__array_function__`.
- Decorator automatic insertion of a `array_only=np._NoValue` (or `coerce` and perhaps `subok=...` if not present) in the function signature, so that users who know that they have arrays only could pass `array_only=True` (name to be decided). This would be most useful if there were also some type of configuration parameter that could set the default of `array_only`.

File: doc/neps/nep-0016-array-function-protocol.rst (hunk 1 above)
Comment (mrocklin) [reply doc/neps/nep-0016-array-function-protocol.rst]: Done

File: doc/neps/nep-0016-array-function-protocol.rst (hunk 2 above)
Comment (mrocklin) [reply doc/neps/nep-0016-array-function-protocol.rst]: Done

Comment (shoyer) [reply doc/neps/nep-0016-array-function-protocol.rst]: I pushed a few commits to your branch already, you probably want to pull them before making further edits :)

File: doc/neps/nep-00XX-array-function-protocol.rst (hunk 21 above)
Comment (mrocklin) [reply doc/neps/nep-00XX-array-function-protocol.rst]: I'm not sure how best to weave this text into the NEP.  To me this seems like orthogonal functionality.

**Comment by mrocklin** [comment]
> rebase against master to remove the failing test
> rename the file to doc/neps/nep-0018-array-function-protocol.rst

Done!   My apologies for the delays on this @mattip .

File: doc/neps/nep-00XX-array-function-protocol.rst (hunk 21 above)
Comment (mhvk) [reply doc/neps/nep-00XX-array-function-protocol.rst]: My text may have been confusing, but I think `__array_function__` is definitely *very* relevant for subclasses. 

For instance, currently `ma.extras` is full of re-definitions of functions like `stack` - i.e., it follows the make-your-own-namespace "solution" that this NEP would like to address. This could be completely replaced by the `__array_function__` proposed here!

Similarly, for `Quantity` we are currently relying on overrides or hope that a function uses ufuncs and does not do `asarray` - and we have no good solution for functions like `concatenate` (not being willing to provide our own namespace...). With `__array_function__`, this would be solved. For instance, for `concatenate` we can ensure that all inputs are converted to the same unit before doing the actual concatenation.

In other words, I probably should have written what looks much like this NEP for `MaskedArray` and `Quantity` alone - hence the request to mention them!

Aside: For both the above, I see possible implementations very much in terms of "1. prepare input arrays; 2. call original function on those; 3. set some properties on outputs" - step (2) is most logically implemented with a `super()` call, which is why I strongly suggest there be an `ndarray.__array_function__`.

**Comment by mattip** [comment]
Thanks @mrocklin

**Comment by shoyer** [comment]
I've now posted the NEP to the mailing list for discussion
https://mail.python.org/pipermail/numpy-discussion/2018-June/078127.html

@mhvk with regards to `ndarray.__array_function__` in particular, I don't have strong feelings here and am happy to differ to your judgment (you know I don't like subclassing). Certainly there is virtue in using the same approach as `__array_ufunc__`. I'll add a brief note on this in the next revision.

With regards to namespaces, please note the section on "Separate namespace" that is already in the NEP text.

File: doc/neps/nep-0018-array-function-protocol. [hunk 23]
```diff
@@ -0,0 +1,543 @@
...
+
+This is not a new problem for NumPy. NumPy has occasionally changed the
+signature for functions in the past, including functions like
+``numpy.sum`` which support overloads.
+
+For adding new keyword arguments that do not change default behavior, we
+would only include these as keyword arguments when they have changed
+from default values. This is similar to `what NumPy already has
+done <https://github.com/numpy/numpy/blob/v1.14.2/numpy/core/fromnumeric.py#L1865-L1867>`__,
+e.g., for the optional ``keepdims`` argument in ``sum``:
```
Comment (mhvk) [inline doc/neps/nep-0018-array-function-protocol.]: Might be worth mentioning that the dance decorator could do this part as well!

**Review by mhvk — COMMENTED** [review]
Two smaller comments; more general ones on the mailing list.

File: doc/neps/nep-0018-array-function-protocol. [hunk 24]
```diff
@@ -0,0 +1,543 @@
...
+within every relevant Numpy function. This is a pervasive change, but of
+fairly simple and innocuous code that should complete quickly and
+without effect if no arguments implement the ``__array_function__``
+protocol. Let us consider a few examples of NumPy functions and how they
+might be affected by this change:
+
+.. code:: python
+
+    def broadcast_to(array, shape, subok=False):
+        success, value = do_array_function_dance(
```
Comment (mhvk) [inline doc/neps/nep-0018-array-function-protocol.]: It's a mini-thing, but one might as well have the dance return either a result or `NotImplemented`.

**Comment by mattip** [comment]
Are we ready to start the approval process for NEP 18?

File: doc/neps/nep-0018-array-function-protocol. (hunk 24 above)
Comment (hameerabbasi) [reply doc/neps/nep-0018-array-function-protocol.]: I’d be in favor of this. No need to complicate things with multiple out args, remember the order and so on.

**Comment by shoyer** [comment]
@mattip almost -- I'd like to add short paragraphs on class methods and dtypes to "Alternatives" to make it clear that we aren't aren't ruling out such options in the future.

**Comment by okuta** [comment]
Hello, I am a [CuPy](https://github.com/cupy/cupy) developer.
This feature is important for NumPy compatible library.
I create a PR for CuPy in experimental cupy/cupy#1650 . Please use it if you want.
Thanks.
